_posts/2022-10-25-using-h3-hexes.md

```diff
@@ -0,0 +1,190 @@
+
+For example, I used `%timeit` to  assign a resolution 9 ID to all 1,339,635 of our [address points]({% link data/location/address-data/index.html %}) (after projecting them to WGS84 to get direct access to lat/long in the SHAPE field):
+
+```python
+def assign_h3(df, resolution):
+    df[f'h3_{resolution}'] = df.apply(lambda row: h3.h3_to_string(h3.geo_to_h3(row['SHAPE']['y'], row['SHAPE']['x'], resolution)), axis=1)
+
+addr_df = pd.DataFrame.spatial.from_featureclass(r'C:\gis\Projects\H3\H3.gdb\address_points_20220727_wgs84')
+
+%timeit assign_h3(addr_df, 9)
+```
+
+`> 23.8 s ± 377 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`
+
+24 Seconds for 1.3 million points. That's fast 🏃‍♀️💨
+
+Note: To get this speed, you'll need to use the `h3.api.numpy_int` API as noted in the `h3-py` [documentation](https://uber.github.io/h3-py/api_comparison)
+
+## Using H3 Hexes in A Map: Analyzing Supermarket Availability
+{: .text-left}
+
+Ok, so we can assign a hex id to point data quickly and easily. But, at some point, we need a spatial representation of the hexes if we're going to map out our summarized or aggregated data. Once we've got that, we can then use pandas to aggregate our data and join it to the hex geometries.
+
+To show this in action, we'll use our [Open Source Places]({% link data/society/open-source-places/index.html %}) and the address points we investigated earlier to do a quick analysis of the number of addresses per grocery store throughout the state.
+
+### Creating an H3 Hex Feature Class
+{: .text-left}
+
+First, we need the actual polygons of the hexes themselves. The H3 API provides the `polyfill` method for identifying all hex IDs whose _centroids_ are within a polygon along with the `h3_to_geo_boundary` method for returning the polygon corresponding to a specific hex ID.
+
+In the code below we first buffer the state boundary so that we get all the hexes that cover the state. Then, we get the hex ids, get the polygons associated with those ids as geojson, and convert the geojson to a spatially-enabled dataframe. Finally, we write that dataframe out to disk.
+
+```python
+state_boundary_df = pd.DataFrame.spatial.from_featureclass(r'C:\gis\Projects\H3\opensgid.agrc.utah.gov.sde\opensgid.boundaries.state_boundary')
+
+#: Buffer the state boundary by 5km
+#: Row index 0 is a mask outside the state boundary, index 1 is the boundary itself.
+#: Also, the SHAPE field is the last item in the column index
+buffered = state_boundary_df.iloc[1, -1].buffer(5000)
+buffered_df = pd.DataFrame.spatial.from_df(
+                pd.DataFrame({'SHAPE': [buffered]}), 
+                geometry_column='SHAPE', sr=26912
+                )
+
+#: Project our polygon to WGS84 and convert to geojson for the H3 analysis
+buffered_df.spatial.project(4326)
```

Comment:
**question**: does the project command allow you to use a transformation?

Reply:
It does, but there's apparently a bug in the way the ArcGIS API for Python calls the underlying functions that causes a serious time disadvantage, making it impossible to project anything more than a very small spatial dataframe. So I'm just ignoring it here.